shiny/ui/_theme_brand.py

```diff
@@ -0,0 +1,410 @@
+
+        # Initialize theme ------------------------------------------------------------
+        super().__init__(
+            name=name,
+            preset=brand_bootstrap.preset,
+            include_paths=include_paths,
+        )
+        self.brand = brand
+
+        # brand.color -----------------------------------------------------------------
+        sass_vars_colors: dict[str, str] = {}
+        sass_vars_brand_colors: dict[str, str] = {}
+        css_vars_brand_colors: list[str] = []
+
+        raise_for_unmapped_vars = (
+            os.environ.get("SHINY_BRAND_YML_RAISE_UNMAPPED") == "true"
+        )
+
+        if brand.color:
+            # Map values in colors to their Sass variable counterparts
+            for thm_name, thm_color in brand.color.to_dict(include="theme").items():
+                if thm_name not in color_map:
+                    if raise_for_unmapped_vars:
+                        raise ThemeBrandUnmappedFieldError(f"color.{thm_name}")
+                    continue
+
+                for sass_var in color_map[thm_name]:
+                    sass_vars_colors[sass_var] = thm_color
+
+            brand_color_palette = brand.color.to_dict(include="palette")
+
+            # Map the brand color palette to Bootstrap's named colors, e.g. $red, $blue.
+            for pal_name, pal_color in brand_color_palette.items():
+                if pal_name in bootstrap_colors:
+                    sass_vars_colors[pal_name] = pal_color
+
+                # Create Sass and CSS variables for the brand color palette
+                color_var = sanitize_sass_var_name(pal_name)
+
+                # => Sass var: `$brand-{name}: {value}`
+                sass_vars_brand_colors.update({f"brand-{color_var}": pal_color})
+                # => CSS var: `--brand-{name}: {value}`
+                css_vars_brand_colors.append(f"--brand-{color_var}: {pal_color};")
+
+        # brand.typography ------------------------------------------------------------
+        sass_vars_typography: dict[str, str] = {}
+        if brand.typography:
+            brand_typography = brand.typography.model_dump(
+                exclude={"fonts"},
+                exclude_none=True,
+            )
+
+            for typ_field, typ_value in brand_typography.items():
+                if typ_field not in typography_map:
+                    if raise_for_unmapped_vars:
+                        raise ThemeBrandUnmappedFieldError(f"typography.{typ_field}")
+                    continue
+
+                for typ_field_key, typ_field_value in typ_value.items():
+                    if typ_field_key in typography_map[typ_field]:
+                        typo_sass_var = typography_map[typ_field][typ_field_key]
+
+                        if typ_field == "base" and typ_field_key == "size":
+                            typ_field_value = str(
+                                maybe_convert_font_size_to_rem(typ_field_value)
+                            )
```

Comment:
The problem of doing calculation with incompatible css units feels more general than this particular Bootstrap problem. Perhaps `Brand` could gain a method for translating CSS units?

Reply:
Yeah I'll think about this. It's currently only relevant for the base font size and Quarto does something different because allowing any unit is important to them. So the key need here is to translate the base font size to rem (and not generally doing css unit translations). But if there's an easy place to hang this in brand I'll put it there

Reply:
Okay, this now happens in `brand_yml` via Pydantic's [serialization context](https://docs.pydantic.dev/2.9/concepts/serialization/#serialization-context). The feature is documented in brand_yml's [`BrandTypographyBase` class docs](https://posit-dev.github.io/brand-yml/pkg/py/typography.html#brand_yml.typography.BrandTypographyBase).

That said, I think we should (in the future) consider using `$font-size-root` instead of `$font-size-base`. First of all, it doesn't have the same incompatible units problem, and second it better scales the entire UI. Both Quarto and bslib [use `$font-size-base` as the font-scaling unit](https://github.com/rstudio/bslib/blob/85b042832e1520b5f06dc95c769396816be76eca/R/bs-theme.R#L240-L242), so it's a good choice for now but something we can revisit. I left a todo note for us in py-shiny.